k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -53,6 +61,58 @@ type DevicePoolStatus struct {
 	AvailableDevices int `json:"availableDevices,omitempty"`
 }
 
+// Device is used to track individual devices in a pool.
+type Device struct {
+	// Name is a driver-specific identifier for the device.
+	// +required
+	Name string `json:"name"`
+
+	// Attributes contain additional metadata that can be used in
+	// constraints. If an attribute name overlaps with the pool attribute,
+	// the device attribute takes precedence.
+	//
+	// +optional
+	Attributes []Attribute `json:"attributes,omitempty"`
+
+	// Requests contains the pooled resources that are consumed when
+	// this device is allocated.
+	//
+	// +optional
+	Requests map[string]resource.Quantity `json:"requests,omitempty"`
```

Comment:
As a hardware vendor, it feels weird to call this requests. I am not requesting anything -- I am declaring what resources my device has that will eventually be requested / consumed by the user.

Reply:
Also, what is the stance on using a `map[string]<type>` in an API object? I was under the impression this was discouraged / not allowed.

Reply:
Also, I feel like we need a level of indirection here rather than just having a `map[string]resource.Quantity` (we might want more than just quantities in the future).

Reply:
Thinking through the implications of this for the end user. If they want a MIG device with a certain placement, they will have to write something in their CEL expression like :
```
device.mig-profile == "3g.20gb" && device.memory-slices-00 == 1 && device.memory-slices-01 == 1 && device.memory-slices-02 == 1 && device.memory-slices-03 == 1
```
rather than:
```
device.mig-profile == "3g.20gb" && device.memory-slices == "0-3"
```

Though maybe the better way to give users control of this would be to create an "index" attribute so that you could do the following to ensure that you get the first "3g.20gb" MIG device on the GPU:
```
device.mig-profile == "3g.20gb" && device.index == "0"
```

Reply:
> As a hardware vendor, it feels weird to call this requests. I am not requesting anything -- I am declaring what resources my device has that will eventually be requested / consumed by the user.

Yeah. Just trying to keep consistent naming across the API for something that "draws down on capacity".

> Also, what is the stance on using a `map[string]<type>` in an API object? I was under the impression this was discouraged / not allowed.

Generally, not allowed as I understand it. But I am keeping consistency with the current way requests are done in PodSpec. I do this here and in the claim. It will be up to the API reviewers to decide which is better - consistency with existing APIs or avoiding the `map[string]<type>`.

> Though maybe the better way to give users control of this would be to create an "index" attribute so that you could do the following to ensure that you get the first "3g.20gb" MIG device on the GPU:

That's an interesting idea, that could work.